packages/compose-tunnel-agent/src/ssh/tunnel-client.ts

```diff
@@ -131,26 +125,35 @@ export const sshClient = async ({
     clientId, tunnels: tunnelsFromHelloResponse(tunnels),
   })
 
+  const getForwardRequestId = (service: {name: string; project: string; port: number; access: 'private' | 'public'}) => {
+    const { tunnel: tunnelName } = tunnelNameResolver({ ...service })
+    const metadata = service.access === 'private' ? '#access=private' : ''
+    return `/${tunnelName}${metadata}`as const
+  }
   let state: SshState
   const limit = plimit(1)
 
   return {
     updateTunnels: async (services: RunningService[]): Promise<SshState> => limit(async () => {
-      const forwards = new Map(
-        services.flatMap(service => service.ports.map(x => tunnelNameResolver({ ...service, port: x }))
-          .map(({ port, tunnel }) => ([tunnel, { host: service.name, port, service }])))
-      )
-
-      const inserts = maps.difference(forwards, existingForwards)
-      const deletes = maps.difference(existingForwards, forwards)
-
-      const haveChanges = inserts.size > 0 || deletes.size > 0
+      const forwards = tunnelSet()
+      services.flatMap(service =>
+        service.ports.map(x => ({ forwardRequestId: getForwardRequestId({ ...service, port: x }), port: x }))
+          .forEach(({ port, forwardRequestId }) => forwards.add(
+            { host: service.name, port, service, sockets: new Set(), forwardRequestId }
+          )))
+
+      const deletes = existingForwards.difference(forwards)
```

Comment:
`inserts` and `deletes` can both be returned from `.difference`

Reply:
there is an importence to doing them one after the other, because we first delete, and then check to see what needs to be inserted. wouldn't work otherwise..